### DISCUSSION TITLE
How can I "trace()" to ESP32 default serial console?

### DISCUSSION THREAD

**Discussion by linfan68** [Moddable-OpenSource/moddable#1077] [Q&A] [answered]
Quick question:
Is there a way to send "trace" to ESP32 default console serial (where printf("") goes), without debugger or debugging build?

**Comment by phoddie** [comment, marked as answer]
There are three kinds of builds:

- debug – for JavaScript debugging with xsbug
- release – for production (serial console disabled, watch dog timer enabled, etc)
- instrumented – a release build with some differences to help with debugging. For example, serial console is enabled, so `trace` statements are output to the serial console.

To run an instrumented "helloworld", replace "-d" with "-i" on the mcconfig command line.

```
cd $MODDABLE/examples/helloworld
mcconfig -i -m -p esp32
```

After installing, the IDF serial monitor is launched, so you see the usual boot sequence followed by the Moddable SDK instrumentation initialization message, and then the usual trace from helloworld.

```
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
...
I (68) heap_init: Initializing. RAM available for dynamic allocation:
I (74) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (79) heap_init: At 3FFB3630 len 0002C9D0 (178 KiB): DRAM
I (84) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (89) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (95) heap_init: At 4008B900 len 00014700 (81 KiB): IRAM
I (100) spi_flash: detected chip: generic
I (104) spi_flash: flash io: qio
I (10instruments key: Pixels drawn,Frames drawn,Network bytes read,Network bytes written,Network sockets,Timers,Files,Poco display list used,Piu command List used,SPI flash erases,Turns,System bytes free,CPU 0,CPU 1,Chunk used,Chunk available,Slot used,Slot available,Stack used,Stack available,Garbage collections,Keys used,Modules loaded,Parser used,Floating Point
Hello, world - sample

**Reply by linfan68** [reply]
Thank you so much!
A follow up question: How can I view the console log from the simulator?

**Reply by phoddie** [reply]
You have a couple of options for console output when running in the simulator. Both require that you run a debug build (passing `-d` to `mcconfig`). By default a debug build launches the xsbug JavaScript debugger along with the simulator. The trace output is displayed in the console pane of xsbug.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/6856458/230136171-c45fd71d-b583-4ac3-94a0-8c8960fcbc1d.png">

Alternatively, you can use xsbug-log which redirects trace output to your console. To use xsbug-log include both `-d -l` when running `mcconfig`.

<img width="949" alt="image" src="https://user-images.githubusercontent.com/6856458/230136520-9a4a25c6-0e73-45ad-aa1a-7f13f1753dfc.png">

> Note: xsbug-log is implemented in Node.js so you will need to do an `npm install` in $MODDABLE/tools/xsbug-log before running it.
